org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -53,33 +56,134 @@
 public class CodeActionHandler {
 
 	public static final String COMMAND_ID_APPLY_EDIT = "java.apply.workspaceEdit";
+	private static final String CODE_ACTION_KIND_QUICK_ASSIST = "quickassist";
 
-	private QuickFixProcessor quickFixProcessor = new QuickFixProcessor();
-
+	private QuickFixProcessor quickFixProcessor;
+	private RefactorProcessor refactorProcessor;
 	private QuickAssistProcessor quickAssistProcessor;
 
+	// TODO (Yan): Remove, move InvertBoolean proposals into RefactorProcessor
 	private AdvancedQuickAssistProcessor advancedQuickAssistProcessor;
 
 	private SourceAssistProcessor sourceAssistProcessor;
 
+
 	private PreferenceManager preferenceManager;
 
 	public CodeActionHandler(PreferenceManager preferenceManager) {
 		this.preferenceManager = preferenceManager;
+		this.quickFixProcessor = new QuickFixProcessor();
 		this.sourceAssistProcessor = new SourceAssistProcessor(preferenceManager);
 		this.quickAssistProcessor = new QuickAssistProcessor(preferenceManager);
 		this.advancedQuickAssistProcessor = new AdvancedQuickAssistProcessor(preferenceManager);
+		this.refactorProcessor = new RefactorProcessor(preferenceManager);
 	}
 
-	/**
-	 * @param params
-	 * @return
-	 */
 	public List<Either<Command, CodeAction>> getCodeActionCommands(CodeActionParams params, IProgressMonitor monitor) {
 		final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
 		if (unit == null) {
 			return Collections.emptyList();
 		}
+
+		int start = DiagnosticsHelper.getStartOffset(unit, params.getRange());
+		int end = DiagnosticsHelper.getEndOffset(unit, params.getRange());
+		InnovationContext context = new InnovationContext(unit, start, end - start);
+		context.setASTRoot(getASTRoot(unit));
+		// TODO (Yan): Problem locations should be calculated only for quickfix code actions.
+		IProblemLocationCore[] locations = this.getProblemLocationCores(unit, params.getContext().getDiagnostics());
+
+		List<String> codeActionKinds = new ArrayList<>();
+		if (params.getContext().getOnly() != null && !params.getContext().getOnly().isEmpty()) {
```

Comment:
Current test case only covers source actions only, add more test cases for other kinds? For example, contextOnly is QuickFix, only calculate quickFix code actions. If contextOnly is blank, calculate all.

Reply:
Sure, will add cases for them

Reply:
added in commit https://github.com/eclipse/eclipse.jdt.ls/pull/1243/commits/34eba8f8e21283ed71c04a8811b0a38cb53b3c80